contribute/design/1.x/csi/20190606-csi-volume-provisioning.md

```diff
@@ -0,0 +1,251 @@
+    - StoragePoolClaim
+    - VolumeControllerImage
+    - VolumeTargetImage
+    - VolumeMonitorImage
+    - ReplicaCount
+    - TargetDir
+    - TargetResourceRequests
+    - TargetResourceLimits
+    - AuxResourceRequests
+    - AuxResourceLimits
+    - RunNamespace
+    - ServiceAccountName
+    - FSType
+    - Lun
+    - ResyncInterval
+    - TargetNodeSelector
+    - TargetTolerations
+  - Volume _(runtime configuration specific to the targeted volume)_
+    - runNamespace
+    - capacity
+    - pvc
+    - isCloneEnable
+    - isRestoreVolume
+    - sourceVolume
+    - storageclass
+    - owner
+    - snapshotName
+- Above configurations can be categorized broadly into two groups:
+  - Volume specific
+  - Kubernetes specific
+
+#### Current Implementation -- Volume Create
+- Uses PVC for below actions:
+  - get .metadata.annotations.volume.kubernetes.io/selected-node
+  - get .metadata.labels.openebs.io/replica-anti-affinity
+  - get .metadata.labels.openebs.io/preferred-replica-anti-affinity
+  - get .metadata.labels.openebs.io/target-affinity
+  - get .metadata.labels.openebs.io/sts-target-affinity
+  - derive statefulset application name
+- Uses CStorPoolList for below actions:
+  - filter CSPs based on StoragePoolClaim name
+  - compares desired volume replica count against available number of pools
+  - maps CSP uid with its .metadata.labels.kubernetes.io/hostname
+- Uses StorageClass for below actions:
+  - get .metadata.resourceVersion
+  - get .metadata.labels.openebs.io/sts-target-affinity
+- Creates a Kubernetes Service
+  - stores its name
+  - stores its cluster IP
+- Creates CStorVolume custom resource
+- Creates Kubernetes deployment for controller
+  - lot of config items are applied here conditionally
+- Create CStorVolumeReplica custom resource
+  - pool selection is done here
+  - lot of config items are applied here conditionally
+  - lot of CVR properties get derived from above config items
+
+#### Current Implementation -- Gaps
+- OpenEBS volumes are aware of higher order resources i.e. kubernetes based PVC & SC
+- In other words OpenEBS volumes cannot be managed without the presence of PVC & SC
+- Inorder to implement CSI as per its standards, OpenEBS volumes should decouple
+itself from being aware of container orchestrator.
+- Logic like pool selection are currently done via go-templating which is not 
+sustainable going forward
+- Building cstor volume controller deployment is currently done via go-templating
+which has become quite complex to maintain
+
+#### Proposed Implementation
+Since OpenEBS itself runs on Kubernetes; limiting OpenEBS provisioning
+to be confined to CSI standards will prove difficult to develop and improve upon
+OpenEBS features.
+
+This proposal tries to adhere to CSI standards and make itself CSI compliant. 
+At the same time, this proposal lets OpenEBS embrace use of Kubernetes custom
+resources and custom controllers to provide storage for stateful applications.
```

Comment:
above sentence and "Inorder to implement CSI as per its standards, OpenEBS volumes should decouple
itself from being aware of container orchestrator." are bit confusing, custom resources and CSI are native to kubernetes, what kind of decoupling we want to achieve?
Will CSI work for other orchestrator's also without or with little modification ?

Reply:
CSI is not native to Kubernetes.
OpenEBS needs to implement CSI on one hand, however since OpenEBS itself runs on Kubernetes, it will depend on Kubernetes features. Due to this dependency current OpenEBS implementation for CSI has deviated from just being compliant with existing methods and payload defined by CSI.
This proposal is an attempt to adhere to CSI standards as well as not limit OpenEBS implement its features (read Kubernetes curstom resources and controllers)

I shall add a FAQ section in this proposal.